### PR TITLE
AppCenter fixes

### PR DESCRIPTION
**Pull request by d2dyno1**
**Resolved / Related Issues**
\-

**Details of Changes**
**Possible** fixes of some App Center reports:
- 3042386929u
- 2267876499u
- 2070687788u
- 152381380u
- 666959963u
- 4026517999u

NOTE: These are not confirmed fixes. This PR **only** bundles potential fixes for above issues.

**Validation**
How did you test these changes?
- [x] Built and ran the app

**Screenshots (optional)**
\-

### CODE REVIEW

File: Files/Views/ColumnShellPage.xaml.cs [hunk 1]
```diff
@@ -698,7 +698,7 @@ private async void KeyboardAccelerator_Invoked(KeyboardAccelerator sender, Keybo
                     break;
 
                 case (false, false, false, _, VirtualKey.F1): // F1, open Files wiki
-                    await Launcher.LaunchUriAsync(new Uri(@"https://files-community.github.io/docs"));
+                    await Launcher.LaunchUriAsync(new Uri(@"https://files-community.github.io/docs")); // TODO update this link for website 2.0
```
Comment (yaichenbaum) [inline Files/Views/ColumnShellPage.xaml.cs:701]: Don't worry about these links for now...

File: Files/UserControls/Widgets/FolderWidget.xaml.cs [hunk 2]
```diff
@@ -140,11 +140,14 @@ public void Dispose()
 
         private async Task GetItemsAddedIcon()
         {
-            foreach (var item in ItemsAdded.ToList())
+            for (int i = 0; i < ItemsAdded.Count; i++)
```
Comment (gave92) [inline Files/UserControls/Widgets/FolderWidget.xaml.cs:143]: The ToList() was added to resolve 2267876499u in 1.9.30. 2267876499u is from 1.9.28, unless the error pops out in 1.9.30 and later I'd leave this as is.

File: Files/Filesystem/FilesystemOperations/Helpers/FilesystemHelpers.cs [hunk 3]
```diff
@@ -824,14 +827,17 @@ public async Task<ReturnResult> MoveItemsFromClipboard(DataPackageView packageVi
...
             var destinations = new List<string>();
-            List<ShellFileItem> binItems = null;
             foreach (var item in source)
             {
                 if (recycleBinHelpers.IsPathUnderRecycleBin(item.Path))
                 {
-                    binItems ??= await recycleBinHelpers.EnumerateRecycleBin();
-                    var matchingItem = binItems.FirstOrDefault(x => x.RecyclePath == item.Path); // Get original file name
-                    destinations.Add(PathNormalization.Combine(destination, matchingItem?.FileName ?? item.Name));
+                    List<ShellFileItem> binItems = await recycleBinHelpers.EnumerateRecycleBin();
```
Comment (gave92) [inline Files/Filesystem/FilesystemOperations/Helpers/FilesystemHelpers.cs:834]: I agree with the null check below but `EnumerateRecycleBin` really shouldn't be put inside the loop.

File: Files/Filesystem/FilesystemOperations/Helpers/FilesystemHelpers.cs [hunk 4]
```diff
@@ -598,19 +598,22 @@ public async Task<ReturnResult> CopyItemsFromClipboard(DataPackageView packageVi
...
-                List<ShellFileItem> binItems = null;
+                List<string> destinations = new List<string>();
                 foreach (var item in source)
                 {
                     if (recycleBinHelpers.IsPathUnderRecycleBin(item.Path))
                     {
-                        binItems ??= await recycleBinHelpers.EnumerateRecycleBin();
-                        var matchingItem = binItems.FirstOrDefault(x => x.RecyclePath == item.Path); // Get original file name
-                        destinations.Add(PathNormalization.Combine(destination, matchingItem?.FileName ?? item.Name));
+                        List<ShellFileItem> binItems = await recycleBinHelpers.EnumerateRecycleBin();
```
Comment (gave92) [inline Files/Filesystem/FilesystemOperations/Helpers/FilesystemHelpers.cs:610]: I agree with the null check below but EnumerateRecycleBin really shouldn't be put inside the loop.

File: Files/Extensions/LinqExtensions.cs [hunk 5]
```diff
@@ -1,11 +1,46 @@
...
+        /// This function is faster than enumerable.Count == 0 since it'll only iterate one element instead of all elements.
+        /// <br/>
+        /// This function is null-safe.
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        /// <param name="enumerable"></param>
+        /// <returns></returns>
+        public static bool IsEmpty<T>(this IEnumerable<T> enumerable) => enumerable == null || !enumerable.Any();
+
+        public static T PreventNull<T>(this T element, T defaultValue, bool throwOnNullDefaultValue = true)
```
Comment (gave92) [inline Files/Extensions/LinqExtensions.cs:24]: This does about the same as `element ?? defaultValue` right?

File: Files/Filesystem/FilesystemOperations/Helpers/FilesystemHelpers.cs (hunk 3 above)
Comment (d2dyno1) [reply Files/Filesystem/FilesystemOperations/Helpers/FilesystemHelpers.cs]: Thanks! A little oversight from me

File: Files/UserControls/Widgets/FolderWidget.xaml.cs (hunk 2 above)
Comment (d2dyno1) [reply Files/UserControls/Widgets/FolderWidget.xaml.cs]: Reverted and added try-catch

File: Files/Filesystem/FilesystemOperations/Helpers/FilesystemHelpers.cs [hunk 6]
```diff
@@ -602,13 +602,12 @@ public async Task<ReturnResult> CopyItemsFromClipboard(DataPackageView packageVi
             {
                 ReturnResult returnStatus = ReturnResult.InProgress;
 
-                List<string> destinations = new List<string>();
+                var destinations = new List<string>();
+                List<ShellFileItem> binItems = await recycleBinHelpers.EnumerateRecycleBin();
```
Comment (gave92) [inline Files/Filesystem/FilesystemOperations/Helpers/FilesystemHelpers.cs:606]: Nope xD Now it queries the recycle bin even if no items are under the recyclebin

Comment (d2dyno1) [reply Files/Filesystem/FilesystemOperations/Helpers/FilesystemHelpers.cs]: 😅 I reverted to the original keeping the fix

**Comment by yaichenbaum** [comment]
@d2dyno1 can you resolve the merge conflicts?

File: Files/BaseLayout.cs [hunk 7]
```diff
@@ -341,22 +341,21 @@ protected IEnumerable<ListedItem> GetAllItems()
...
         {
-            IEnumerable items = GetAllItems();
+            IEnumerable<ListedItem> items = GetAllItems();
             if (items == null)
             {
                 return;
             }
 
-            foreach (ListedItem listedItem in items)
+            for (int i = 0; i < items.Count(); i++)
```
Comment (hez2010) [inline Files/BaseLayout.cs:350]: This would result in count calculation for each loop iteration. You may want to call `ToList()` first.

**Review by hez2010 — CHANGES_REQUESTED** [review]

**Review by winston-de — APPROVED** [review]

**Pull request merged**